tendrl/gluster_integration/sds_sync/__init__.py

```diff
@@ -55,6 +55,235 @@ def _emit_event(self, resource, curr_value, msg, instance):
             )
         )
 
+    def _sync_cluster_utilization_details(self, volumes):
+        cluster_raw_capacity = 0
+        cluster_used_capacity = 0
+        cluster_usable_capacity = 0
+        for volume in volumes:
+            subvol_count = 0
+            bricks = []
+            raw_capacity = 0
+            raw_used = 0
+            up_bricks = 0
+            subvols = []
+            while True:
+                try:
+                    subvol = NS._int.client.read(
+                        "clusters/%s/Volumes/%s/Bricks/subvolume%s" % (
+                            NS.tendrl_context.integration_id,
+                            volume.vol_id,
+                            subvol_count
+                        )
+                    )
+                    subvol_bricks = []
+                    for entry in subvol.leaves:
+                        brick_name = entry.key.split("/")[-1]
+                        fetched_brick = NS.gluster.objects.Brick(
+                            name=brick_name
+                        ).load()
+                        raw_capacity += fetched_brick.utilization['total']
+                        cluster_raw_capacity += fetched_brick.utilization['total']
+                        raw_used += fetched_brick.utilization['used']
+                        cluster_used_capacity += fetched_brick.utilization['used']
+                        bricks.append(fetched_brick)
+                        subvol_bricks.append(fetched_brick)
+                        if fetched_brick.status != "Started":
+                            up_bricks += 1
+                    subvols.append(subvol_bricks)
+                    subvol_count += 1
+                except etcd.EtcdKeyNotFound:
+                    break
+            vol_usable_capacity = 0
+            vol_used_capacity = 0
+            vol_pcnt_used = 0
+            if up_bricks == len(bricks):
+                if int(volume.disperse_count) == 0:
+                    # This is distribute or replicate volume
+                    vol_usable_capacity = raw_capacity / int(volume.replica_count)
+                    vol_used_capacity = raw_used / int(volume.replica_count)
+                else:
+                    # this is a disperse volume, with all bricks online
+                    # assumption : all bricks are the same size
+                    disperse_yield = \
+                        float(
+                            int(volume.disperse_count) - int(volume.redundancy_count)
+                        ) / int(volume.disperse_count)
+                    vol_usable_capacity = raw_capacity * disperse_yield
+                    vol_used_capacity = raw_used * disperse_yield
+            else:
+                if int(volume.replica_count) > 1 or int(volume.disperse_count) > 0:
```

Comment:
I feel, for dispersed volume the below logic might be incorrect as we are not considering the redundancy factor at all